### ISSUE TITLE
notebook.source.organizeImports setting does not seem to work when imports are below markdown cell

### ISSUE THREAD

**Issue by bolliger32** [astral-sh/ruff-vscode#593]
It took me a little while, but I finally figured out the use of `notebook.source.fixAll` and `notebook.source.organizeImports` settings (instead of their `source.*` counterparts) to prevent vscode from turning my import cells in jupyter notebooks into gibberish when using ruff as the formatter. However, it appears that when a notebook starts with a markdown cell (say, to title the notebook) and then has a python cell with imports after the markdown, `notebook.source.organizeImports` doesn't work. Here are my relevant vscode settings:

```json
{
    "editor.defaultFormatter": "charliermarsh.ruff",
    "editor.formatOnSave": true,
    "editor.codeActionsOnSave": {
        "source.organizeImports": "explicit",
        "source.fixAll": "explicit"
    },
    "notebook.defaultFormatter": "charliermarsh.ruff",
    "notebook.formatOnSave.enabled": true,
    "notebook.codeActionsOnSave": {
        "notebook.source.organizeImports": "explicit",
        "notebook.source.fixAll": "explicit"
    }
}
```

And here's the relevant part of my `pyproject.toml` file in the workspace open in vscode:

```toml
[tool.ruff.lint]
unfixable = ["F401"]
select = ["E4", "E7", "E9", "F", "NPY", "I"]

[tool.ruff.lint.per-file-ignores]
"__init__.py" = ["E402"]

[tool.ruff.format]
docstring-code-format = true
```

I'm using ruff 0.6.2.

Side Note:
If I use `source.organizeImports` instead of `notebook.source.organizeImports`, it "works" in the sense that something happens, but the effect is to introduce gibberish. Here's an example of the "gibberish" that occurs when using `source.*` instead of `notebook.source.*`. The first screenshot is from "pre-save", the second is after saving once, and the third after saving a second time
<img width="343" alt="image" src="https://github.com/user-attachments/assets/8bf52b42-f6b3-44a0-b1ab-0e40d31b88d6">
<img width="311" alt="image" src="https://github.com/user-attachments/assets/bed7e85b-7630-431a-bf51-44a8de2a4448">
<img width="298" alt="image" src="https://github.com/user-attachments/assets/2a1b72a6-e286-4137-a7d6-21db5544f014">

**Comment by dhruvmanila** [comment]
Huh, that is indeed the case. I wonder if it's a bug in VS Code, let me check if the editor is sending the request in the first place.

**Comment by dhruvmanila** [comment]
Hmm, VS Code isn't sending the code action request when a markdown cell is present. This is a bit worrying and could be related to the notebook selector filter here: https://github.com/astral-sh/ruff/blob/3e9c7adeeee639043686a1717a1683f6620efd2a/crates/ruff_server/src/server.rs#L335-L344

**Comment by dhruvmanila** [comment]
Let me look at this later today. Thanks for raising this issue!

**Comment by bolliger32** [comment]
Thanks for taking a look so quickly!

**Comment by dhruvmanila** [comment]
This _might_ be related to https://github.com/microsoft/vscode/issues/216653

**Comment by Riezebos** [comment]
@bolliger32 Thanks for mentioning `"notebook.source.organizeImports"` as a replacement for `"source.organizeImports"`! I had no idea this existed, and since a few weeks I've had a bunch of import cells in notebooks turn to gibberish as well. It was quite annoying as the Undo action would also not consistently restore the original version.

**Comment by matt-dies-tenet3** [comment]
I commented on the potentially related issue as well, but I'm experiencing the same issue; manually saving kicks off the importing, but doesn't actually organize the imports and instead creates gibberish.

Unfortunately, the VSCode issue doesn't seemed to have gained any traction quite yet, but fingers crossed!

**Comment by dhruvmanila** [comment]
@Riezebos Does your issue gets resolved when using `notebook.source.organizeImports` instead of `source.organizeImports`?

@matt-dies-tenet3 Can you try doing the same? Use `notebook.source.organizeImports.ruff` instead of `source.organizeImports`?

**Comment by Riezebos** [comment]
@dhruvmanila yes, using `notebook.source.organizeImports` instead of `source.organizeImports` inside of `"notebook.codeActionsOnSave"` stopped imports from becoming gibberish.

**Comment by MichaReiser** [comment]
Has the VS code extension access to the `notebook.codeActionsOnSaveSetting` so that we could show a warning / info message and recommend users to use `notebook.source.organizeImports` instead?

**Comment by dhruvmanila** [comment]
> Has the VS code extension access to the `notebook.codeActionsOnSaveSetting` so that we could show a warning / info message and recommend users to use `notebook.source.organizeImports` instead?

Yes, it should. Although I want to look into why `ruff-lsp` isn't facing the same issue.

**Comment by dhruvmanila** [comment]
Ok, I think I've narrowed this down to why it's happening in the native server and not `ruff-lsp` for `source.organizeImports.ruff`:
- `ruff-lsp` receives the request for each individual cells and sends the response for that specific cell only
- Native server receives the request for each individual cells and sends the response for the entire notebook which means it sends two edit request to the client which results in gibberish text

I think the reason is that Ruff requires context of the entire notebook but that's only relevant for `fixAll` and not `organizeImports` (import organization doesn't require context from other cells). So, the solution would be to:
1. Remove support for non-notebook prefixed code actions from native server
2. Fix this issue such that applying code actions on individual cells work

I'm leaning towards (1) because (2) will still be problematic for the "fixAll" code action and will only work for "organizeImports" code action.

**Comment by matt-dies-tenet3** [comment]
Using the following also works consistently!

```json
"notebook.formatOnSave.enabled": true,
"notebook.codeActionsOnSave": {
    "notebook.source.organizeImports.ruff": "explicit"
},
```

**Comment by dhruvmanila** [comment]
Oh wow, I just found out this:

> note: if you are linking your provider to a selector that only runs against code cells, this will likely break functionality, as markdown cells are often the leading cell in notebooks. This behavior will be explored during next debt week.
>
> https://github.com/microsoft/vscode/issues/179213#issuecomment-1571065906

which is basically what is happening here.

**Comment by dhruvmanila** [comment]
cc @Yoyokrazy, apologies for the ping but if possible could you clarify if this was discussed further as I couldn't find any related issues. 

We're facing this issue here as we do select only Python code cells during initialization:

https://github.com/astral-sh/ruff/blob/6e34f74c164a646f07f3fe9bed9492c3557a95d7/crates/ruff_server/src/server.rs#L320-L329

**Comment by dhruvmanila** [comment]
> Even when manually triggered by command, it still says no available imports organizer found. In the meanwhile, the formatting/linting features of ruff work well as normal in jupyter. And in regular python files, the imports organizer could be found.

IIUC, this seems unrelated to what's being discussed here. Do you mind opening a new issue with all the details (refer to [troubleshooting guide](https://github.com/astral-sh/ruff-vscode#troubleshooting)).

**Comment by dhruvmanila** [comment]
I might try changing the notebook selector to include markdown cells on syncing, I think the server should be able to handle it.

**Comment by dhruvmanila** [comment]
> I might try changing the notebook selector to include markdown cells on syncing, I think the server should be able to handle it.

Nope, this doesn't seem to work either. I've tried the following [`NotebookDocumentSyncOptions`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notebookDocumentSyncOptions) but none of them allows VS Code to send us the notebook source code action request but when I rearrange cells such that the first cell is a code cell then it works:

```json
        "notebookDocumentSync": {
            "notebookSelector": [
                {
                    "cells": [
                        {
                            "language": "python"
                        },
                        {
                            "language": "markdown"
                        }
                    ],
                    "notebook": {
                        "notebook_type": "jupyter-notebook"
                    }
                }
            ],
            "save": false
        },
```

```json
        "notebookDocumentSync": {
            "notebookSelector": [
                {
                    "cells": [
                        {
                            "language": "python"
                        },
                        {
                            "language": "markdown"
                        }
                    ]
                }
            ],
            "save": false
        },
```

```
        "notebookDocumentSync": {
            "notebookSelector": [
                {
                    "cells": [
                        {
                            "language": "*"
                        }
                    ]
                }
            ],
            "save": false
        },
```

**Comment by QuentinSoubeyranAqemia** [comment]
I wonder if comment that the command `Organize Imports` doesn't work with ruff server might be relevant? There are two command to sort the imports, VSCode's `Organize Imports` and ruff's `Ruff: Format imports`.

The old `ruff-lsp` supports both, but the new `ruff server` doesn't support the `Organize Imports` (VScode shows a message "no organize import action available"). Could one of the reason that ruff isn't getting the event from VSCode be because VSCode thinks ruff cannot handle the `notebook.source.organizeImports` action?

I made a detailed comparison of ruff lsp and ruff server below with the details. I'm using ruff `0.10.0` bundled with VSCode extension version `2025.16.0`.

----

<details>
<summary>Click to expand detailed configuration & results</summary>

Here's what I tried. The test notebooks contains up to two cells: an optional markdown title and a python cell. 
```markdown
# test
```

```python

import sys, pathlib
```

What I want to achieve is that upon saving the notebook, the python cell is formatted and sorted, yielding

```python
import pathlib
import sys
```

In the following tables, I use:
- "format" to mean the leading newline is removed
- "imports" to mean the imports are sorted and separated
- "delete" to mean the imports are entirely removed (due to not being used anywhere, ruff fixAll removes them I guess)

Every time I change the configuration, I reload the VSCode window with the command `Developper: Reload Window` to ensure it is properly applied.

## Using `ruff-lsp`


Configuration:
```json
{
  "editor.formatOnSave": true,
  "notebook.codeActionsOnSave": {
        "notebook.source.organizeImports": "explicit",
        "notebook.source.fixAll": "explicit"
    },
    "notebook.defaultFormatter": "charliermarsh.ruff",
    "notebook.formatOnSave.enabled": true,
    "ruff.importStrategy": "useBundled",
    "ruff.nativeServer": "off",
    "ruff.organizeImports": true,
    "ruff.trace.server": "messages",
}
```

| Notebook | save | `Organize Imports` | `Ruff: format imports` |
|-:|:-:|:-:|:-:|
| Leading markdown | format | sort | sort|
| Only python | format+sort+delete | sort | sort | 

## Using `ruff server`


Configuration:
```json
{
  "editor.formatOnSave": true,
  "notebook.codeActionsOnSave": {
        "notebook.source.organizeImports": "explicit",
        "notebook.source.fixAll": "explicit"
    },
    "notebook.defaultFormatter": "charliermarsh.ruff",
    "notebook.formatOnSave.enabled": true,
    "ruff.importStrategy": "useBundled",
    "ruff.nativeServer": "on",
    "ruff.organizeImports": true,
    "ruff.trace.server": "messages",
}
```

| Notebook | save | `Organize Imports` | `Ruff: format imports` |
|-:|:-:|:-:|:-:|
| Leading markdown | format | error | sort |
| Only python | format+sort+delete | error | sort |

## With explicit ruff calls

I try to leverage ruff commands mentioned in https://github.com/astral-sh/ruff/discussions/15991#discussioncomment-12193766 with the following configuraiton:

Configuration:
```json
{
  "editor.formatOnSave": true,
  "notebook.codeActionsOnSave": {
        "notebook.source.organizeImports": "explicit",
        "notebook.source.fixAll": "explicit",
        "ruff.executeOrganizeImports": "explicit",
        "ruff.executeAutoFix": "explicit"
    },
    "notebook.defaultFormatter": "charliermarsh.ruff",
    "notebook.formatOnSave.enabled": true,
    "ruff.importStrategy": "useBundled",
    "ruff.nativeServer": "off", // "on" also tested
    "ruff.organizeImports": true,
    "ruff.trace.server": "messages",
}
```
 It does not change the behavior.

</details>

**Comment by dhruvmanila** [comment]
@QuentinSoubeyranAqemia That's an interesting observation, thanks for sharing that with all the details! I'm not exactly sure what VS Code does when a user invokes the "Organize Imports" command. I think that might give us more clue as to whether this is an issue on the server side or is something that needs to be handled by the editor.

**Comment by dhruvmanila** [comment]
A quick search in the VS Code repository gave me the following code:

https://github.com/microsoft/vscode/blob/420bde3a4c932c519dc4e7b54cf7ac9518eacd4c/src/vs/editor/contrib/codeAction/browser/codeActionCommands.ts#L216-L242

which I think is just invoking the `source.organizeImports` action if it's available otherwise showing the message "No organize imports action available".

**Comment by QuentinSoubeyranAqemia** [comment]
Happy to help. I was trying to get my notebook to work, and this was a good occasion for writing up test results.

I notice `ruff-lsp` doesn't support "markdown"-notebooks either. I might be confused about this, but I seemed to work in the past. I might try to bisect the extension version to see if I'm making things up.

**Comment by QuentinSoubeyranAqemia** [comment]
Dumping some ideas & thoughts here about https://github.com/astral-sh/ruff-vscode/issues/593#issuecomment-2684797904 above:
- are we sure the notebook type is `jupyter-notebook`? Since python-started notebook and markdown-started notebook behave differently, and reading https://github.com/microsoft/vscode/issues/179213#issuecomment-1571065906 again , I wonder if it is actually the type of the first cell?
  - Is there a way to accept anything (`"notebook_type": "*"` perhaps?) in a dev build and log what is sent by VSCode ? I wonder if `"notebook_type": "markdown"` or something similar would work?

- has the `save` parameter been varied? I notice it is `false` in all trials, and I'm not sure what it actually does. I did not search the VSCode documentation, so I'm not sure if this is related or not, but it is a constant in all the non-working tries.

**Comment by totkeks** [comment]
Thanks for this issue. I was going crazy trying to set it to work properly and removing the initial markdown fixed it, even though that is just a bad workaround :D

I also noticed that `organizeImports` is redundant when `fixAll` is configured.

**Comment by mathisdrn** [comment]
I am also having this issue and it took me a while to find this GitHub issue. You could add some note about it in the README : 
```json
 "notebook.codeActionsOnSave": {
    "notebook.source.fixAll": "explicit",
    "notebook.source.organizeImports": "explicit", // doesn't work but is used in doc
    "notebook.organizeImports": "explicit", // doesn't work either
    "source.organizeImports.ruff": "explicit", // doesn't work either
    "source.organizeImports": "explicit", // only this one currently works
  }
```

**Comment by rifatno1** [comment]
Any updates on this issue?

**Comment by MichaReiser** [comment]
This is mostly a VS Code bug. It doesn't send the code action if the first cell is one that doesn't match the LSP's selector 

https://github.com/microsoft/vscode/blob/f09d623b525040809adb8abf4dcc70b228be07ba/src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts#L394-L405

See https://github.com/microsoft/vscode/issues/216653

We could probably work around this on our end by syncing all notebook cells and then filter them before constructing our internal notebook representation.
